src/xia2/Applications/xia2setup.py

```diff
@@ -226,46 +226,87 @@ def visit(directory, files):
+    structure, and may follow a standard for structured (meta)data, such as the NXmx
+    application definition of the NeXus data standard.  In such cases, the top-level
+    file is connected to the subordinate files by HDF5 external links.
+
+    For historical reasons, such a data layout may use two or more top-level files,
+    which are duplicates of each other.  For example, this is common at Diamond Light
+    Source for data from Dectris Eiger detectors, because XDS specifically requires
+    the top-level file to have a name ending in '_master.h5', whereas Diamond's
+    internal standard is for top-level files following the NeXus standard to have a
+    '.nxs' file extension.
+
+    To avoid these duplicate top-level HDF5 files being misidentified as separate
+    sweeps, this function serves to associate such top-level files with the
+    underlying image data sets to which they are linked, and keep only one top-level
+    file for each unique set of image data files.  This identification depends on the
+    image data sets in the top-level file (which may be external links) having names
+    like '/entry/data/data_<stuff>', where <stuff> is usually a string of numerals
+    (see '_linked_hdf5_data_files').
+
+    There are two known weakness of this method:
+      - If one genuinely wishes to import multiple top-level files pointing to the
+        same underlying image data (but perhaps with different metadata), they will
+        be erroneously deduplicated.
+      - The use of image data sets named '/entry/data/data_<stuff>' is not a
+        requirement of a valid top-level HDF5 data file.  The top-level file will
+        always have an image data set named '/entry/data/data', and may not bother
+        with any redundant 'data_<stuff>' data sets/links.  Even if the image data
+        are in separate files to the top-level file, '/entry/data/data' may link to
+        them using the separate HDF5 virtual data set formalism, which obviates the
+        need for external links called 'data_<stuff>'.  In such cases, any duplicates
+        will be missed.
+
+    Args:
+        sweeps:  The sweep data file names, or file name templates.
+
+    Returns:
+        The unique sweep names, with duplicate top-level HDF5 files removed.
+    """
+    deduplicated = set()
+    hdf5_sweeps: dict[frozenset[str], str] = {}
 
     for s in sweeps:
-        if not is_hdf5_name(s):
-            if s not in rest:
-                rest.append(s)
-            continue
-        filenames = tuple(_list_hdf5_data_files(s))
-        if filenames in h5_data_to_sweep:
-            # impose slight bias in favour of using _master.h5 in place of .nxs
-            # because XDS
-            if h5_data_to_sweep[filenames].endswith(".nxs") and s.endswith(
-                "_master.h5"
-            ):
-                h5_data_to_sweep[filenames] = s
+        if not is_hdf5_name(s) or not (filenames := _linked_hdf5_data_files(s)):
```

Comment:
@benjaminhwilliams  superb if sneaky use of walrus operator _however_ ISTR such are verboten until our chums at CCP4 have a platform which supports same? Idiomatically excellent application of same though.

Reply:
Ah, I'd forgotten that walruses are forbidden.  Since CI checks run on 3.8 & 3.9 only, I assumed all distribution was OK.  Are CCP4 really still trying to release on Python < 3.8?